Added support for ReactiveHealthIndicator

joyita07 opened the pull request:
_EurekaHealthCheckHandler_ now considers health status from _ReactiveHealthIndicator_.

pivotal-issuemaster commented:
@joyita07 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3901)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3901) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@joyita07 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3901)!

OlgaMaciaszek reviewed (changes requested):
Since `ReactiveHealthIndicator` is already available in `2.2.x` branch, we should be able to add the changes there. @joyita07 feel free to us the other constructor in the `2.2.x` test instead of the deprecated one. You can switch it for the existing test object or create a new `healthAggregator` instance for your tests.

OlgaMaciaszek commented on spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandler.java:
```diff
@@ -50,7 +51,8 @@
  * @see StatusAggregator
  */
-public class EurekaHealthCheckHandler implements HealthCheckHandler, ApplicationContextAware, InitializingBean {
+public class EurekaHealthCheckHandler
```
Please add your name to class authors with `@author` tag.

OlgaMaciaszek commented on spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandlerTests.java:
```diff
@@ -48,22 +51,23 @@
+		initialize(UpHealthConfiguration.class, ReactiveUpHealthConfiguration.class);
 
-		InstanceStatus status = healthCheckHandler.getStatus(InstanceStatus.UNKNOWN);
+		InstanceStatus status = this.healthCheckHandler.getStatus(InstanceStatus.UNKNOWN);
```
Please remove unnecessary `this.` keywords.

OlgaMaciaszek commented on spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandlerTests.java:
```diff
@@ -72,7 +76,35 @@ public void testDown() throws Exception {
+
+	@Test
+	public void testDown2() throws Exception {
+
```
Please remove the unnecessary empty lines at the beginning of the test. You can add dividing line between the `act` statement, where `status` is retrieved and the assertion.

Please add more meaningful names to tests. For example `testReturnDownWhenBlockingIndicatorsUpAndReactiveDown` is a better name than `testDown2`.

OlgaMaciaszek commented on spring-cloud-netflix-eureka-client/src/test/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandlerTests.java:
```diff
@@ -48,22 +51,23 @@
 	public void setUp() {
 
-		healthCheckHandler = new EurekaHealthCheckHandler(new SimpleStatusAggregator());
+		this.healthCheckHandler = new EurekaHealthCheckHandler(
```
Please add your name to class authors with `@author` tag.

OlgaMaciaszek commented on spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandler.java:
```diff
@@ -104,20 +115,39 @@ void populateHealthIndicators(Map<String, HealthIndicator> healthIndicators) {
+		if (this.reactiveHealthIndicators != null) {
+			statusSet.addAll(this.reactiveHealthIndicators.values().stream()
+					.map(ReactiveHealthIndicator::health).map(health -> {
+						return health.block().getStatus();
```
I would usually suggest using non-blocking calls on reactor objects, but since the entire `EurekaHealthCheckHandler` is not reactive, I think we can leave it this way.

OlgaMaciaszek commented:
Fixes gh-3888.

joyita07 commented on spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaHealthCheckHandler.java:
```diff
@@ -104,20 +115,39 @@ void populateHealthIndicators(Map<String, HealthIndicator> healthIndicators) {
 
 	protected InstanceStatus getHealthStatus() {
-		Status status = getStatus(statusAggregator);
+		Status status = getStatus(this.statusAggregator);
```
I am using the eclipse-code-formatter.xml from the Spring Cloud Build project, not sure if because of that saving the file in Eclipse automatically adds `this.` everywhere. Do you know if there's any way to fix this?

OlgaMaciaszek replied:
Hmmm... this might be deprecated. We usually use IntelliJ. Will remove these manually.

joyita07 commented:
@OlgaMaciaszek I wasn't able to update this PR after making the change in 2.2.x branch, since this one is created against master branch from my forked repo. Created a new [PR](https://github.com/spring-cloud/spring-cloud-netflix/pull/3903) after addressing all your comment. I am closing this one.
